### PR TITLE
Make project compatible with recent python versions

### PR DESCRIPTION
**Pull request by rvhonorato**
You are about to submit a new Pull Request. Before continuing make sure you read the [contributing guidelines](CONTRIBUTING.md).

## Checklist

- [X] Tests added for the new code (new actions)
- [X] Documentation added for the code changes
- [x] Does not break licensing
- [x] Does not add any dependencies, if it does please add a thorough explanation

## Summary of the Pull Request  

This PR makes the project compliant with python 3.9, 3.10, 3.11 and 3.12. Here I implemented the newer `pyproject.toml` in place of the `setup.py` ([which is deprecated for almost 5 years](https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html#summary)). Keep in mind that while this file still exists in the repository it [**must not** be called directly](https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html#summary), I have also updated `DEVELOPMENT.md` to reflect this. This also made it necessary to simplify the custom build class. 

Two notable changes are in `tests/test_libalign.py` and `src/haddock/libs/libalign.py` - please check the files.

The doc build part is very weirdly implemented and it broke with this PR, I will fix it in another.
***

EDIT: I decided to give it another try and refactored the tests to accomodate for the new behaviour of `mkdtemp` introduced in 3.12 - https://github.com/haddocking/haddock3/pull/1058

~I could not make it compatible with 3.12 (the recomended python version) because of a change introduced in [`tempfile.mkdtemp`](https://docs.python.org/3/library/tempfile.html#tempfile.mkdtemp)~

~> [mkdtemp()](https://docs.python.org/3/library/tempfile.html#tempfile.mkdtemp) now always returns an absolute path, even if dir is relative.~

~**A lot** of the tests follow this pattern:~

```python
with tempfile.NamedTemporaryDirectory(dir=".") as tempfile:
    # ...
```

~And changing this to use the appropriate syntax - which would then resolve to `/tmp/` broke all the tests - https://github.com/haddocking/haddock3/pull/1042/commits/d70445a67cc179df499614f962592e5f21f7c569 / https://github.com/haddocking/haddock3/pull/1042/commits/f7e329786550e35e72454c41b72b7287e36fc2df, this can be an indication that haddock3 is tightly coupled with relative paths everywhere. **This needs to be addressed (in the tests) before we can add 3.12.**~
***
As for the dependencies I've managed to do some creative combinations using wildcard selections but these also need to be updated asap. We are currently using very old versions of the dependencies (such as numpy 1.x), this also should be done in a different PR after the relative paths are investigated.

## Related Issue

#1040

### CODE REVIEW

File: src/haddock/libs/libalign.py [hunk 1]
```diff
@@ -1043,23 +1035,23 @@ def postprocess_alignment(self, ref_ch, mod_ch, align_id):
             alignment id (index of the alignment)
         """
         self.align_dic[ref_ch] = {}
-        if not any(e for e in self.top_alns[align_id].aligned):
+        if not np.any(self.top_alns[align_id].aligned):
```
Comment (rvhonorato) [inline src/haddock/libs/libalign.py:1038]: This here was causing 

```
The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

Yes it took me hours to find it 🫠

**Review by mgiulini — CHANGES_REQUESTED** [review]
Looks good but the sphinx commands cannot build the developer documentation (see below)

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -68,6 +55,7 @@ jobs:
           coverage-reports: ./coverage.xml
 
       - name: check if docs are buildable
-        # this step can fail
         continue-on-error: true
-        run: tox -e docs
+        run: |
+          sphinx-apidoc -f -e -o docs/ src/haddock -d 1
+          sphinx-build docs haddock3-docs
```
[the remark this answers is not in the document]
Comment (rvhonorato) [reply .github/workflows/ci.yml:61]: Yep, i said this in the PR description:

> The doc build part is very weirdly implemented and it broke with this PR, I will fix it in another.

File: tests/test_module_rmsdmatrix.py [hunk 3]
```diff
@@ -1,44 +1,37 @@
...
-            score=42.0
-            ),
-        PDBFile(
-            Path(golden_data, "protdna_complex_2.pdb"),
-            path=golden_data,
-            score=28.0
-            )]
+@pytest.fixture(name="rmsdmatrix")
+def fixture_rmsdmatrix():
+    """???"""
```
Comment (mgiulini) [inline tests/test_module_rmsdmatrix.py:21]:
```suggestion
    """rmsdmatrix module fixture"""
```

File: tests/test_module_rmsdmatrix.py [hunk 4]
```diff
@@ -1,44 +1,37 @@
...
+    with tempfile.TemporaryDirectory() as tempdir:
+        os.chdir(tempdir)
+        yield Rmsdmatrix(
+            order=2, path=Path("."), initial_params=DEFAULT_RMSDMATRIX_PARAMS
+        )
+    # rmsd_module.previous_io.output = protdna_input_list
```
Comment (mgiulini) [inline tests/test_module_rmsdmatrix.py:27]:
```suggestion
        )
```

File: tests/test_module_topoaa.py [hunk 5]
```diff
@@ -20,45 +21,36 @@
...
-            initial_params=topoaa_params)
-
-
-@pytest.mark.skip(reason="Not implemented yet")
-def test_confirm_installation(topoaa):
-    # topoaa.confirm_installation()
-    pass
+@pytest.fixture(name="topoaa")
+def fixture_topoaa():
+    """???"""
```
Comment (mgiulini) [inline tests/test_module_topoaa.py:50]:
```suggestion
    """topoaa module fixture"""
```

File: tests/test_module_mdscoring.py [hunk 6]
```diff
@@ -1,55 +1,73 @@
...
-        path=Path("1_emscoring"),
-        initial_params=mdscoring_pars
+from haddock.modules.scoring.mdscoring import \
+    DEFAULT_CONFIG as MDSCORING_DEFAULT_PARAMS
+from haddock.modules.scoring.mdscoring import HaddockModule as MDScoring
+
+
+@pytest.fixture(name="mdscoring")
+def fixture_mdscoring():
+    """???"""
```
Comment (mgiulini) [inline tests/test_module_mdscoring.py:16]:
```suggestion
    """mdscoring module fixture"""
```

File: tests/test_module_mdscoring.py [hunk 7]
```diff
@@ -1,55 +1,73 @@
...
-    mds_module.output_models = output_models
-    for mod in range(len(output_models)):
-        ori_name = "original_name_" + str(mod) + ".pdb"
-        mds_module.output_models[mod].ori_name = ori_name
-    # creating output
+
+
+@pytest.fixture(name="mdscoring_dna")
+def fixture_mdscoring_dna(mdscoring, protdna_input_list):
+    """???"""
```
Comment (mgiulini) [inline tests/test_module_mdscoring.py:27]:
```suggestion
    """mdscoring module fixture protdna"""
```

File: tests/test_module_seletopclusts.py [hunk 8]
```diff
@@ -360,30 +341,34 @@ def test_select_top_sized_clusts_models(clustered_models):
...
     """Test sorting of ranked models."""
     sorted_models, _note = sort_models(ranked_models)
     assert not _note
-    assert sorted_models[0].file_name == "protprot_complex_2.pdb"
-    assert sorted_models[1].file_name == "protprot_complex_1.pdb"
+    assert sorted_models[0].file_name == ranked_models[1].file_name
+    assert sorted_models[1].file_name == ranked_models[0].file_name
+
+    # assert sorted_models[0].file_name == "protprot_complex_2.pdb"
+    # assert sorted_models[1].file_name == "protprot_complex_1.pdb"
```
Comment (mgiulini) [inline tests/test_module_seletopclusts.py:359]: can we remove this? or do you want to keep the commented lines for future reference?

**Review by mgiulini — DISMISSED** [review]

**Review by mgiulini — APPROVED** [review]

**Review by amjjbonvin — APPROVED** [review]
Nice!

Out of curiosity: where are the tests actually running when the path is defined as path=Path(".") ? Under some /var/tmp directory?

**Pull request merged**